p2p/transport/webrtc/connection.go

```diff
@@ -0,0 +1,275 @@
+package libp2pwebrtc
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+
+	ic "github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	tpt "github.com/libp2p/go-libp2p/core/transport"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/pion/webrtc/v3"
+)
+
+var _ tpt.CapableConn = &connection{}
+
+const (
+	maxAcceptQueueLen int = 10
+)
+
+type connection struct {
+	pc        *webrtc.PeerConnection
+	transport *WebRTCTransport
+	scope     network.ConnManagementScope
+
+	localPeer      peer.ID
+	privKey        ic.PrivKey
+	localMultiaddr ma.Multiaddr
+
+	remotePeer      peer.ID
+	remoteKey       ic.PubKey
+	remoteMultiaddr ma.Multiaddr
+
+	m       sync.Mutex
+	streams map[uint16]*dataChannel
+
+	acceptQueue chan network.MuxedStream
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func newConnection(
+	pc *webrtc.PeerConnection,
+	transport *WebRTCTransport,
+	scope network.ConnManagementScope,
+
+	localPeer peer.ID,
+	privKey ic.PrivKey,
+	localMultiaddr ma.Multiaddr,
+
+	remotePeer peer.ID,
+	remoteKey ic.PubKey,
+	remoteMultiaddr ma.Multiaddr,
+) *connection {
+	acceptQueue := make(chan network.MuxedStream, maxAcceptQueueLen)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	conn := &connection{
+		pc:        pc,
+		transport: transport,
+		scope:     scope,
+
+		localPeer:      localPeer,
+		privKey:        privKey,
+		localMultiaddr: localMultiaddr,
+
+		remotePeer:      remotePeer,
+		remoteKey:       remoteKey,
+		remoteMultiaddr: remoteMultiaddr,
+		ctx:             ctx,
+		cancel:          cancel,
+		streams:         make(map[uint16]*dataChannel),
+
+		acceptQueue: acceptQueue,
+	}
+
+	pc.OnDataChannel(func(dc *webrtc.DataChannel) {
+		id := *dc.ID()
+		var stream *dataChannel
+		dc.OnOpen(func() {
+			// datachannel cannot be detached before opening
+			rwc, err := dc.Detach()
+			if err != nil {
+				log.Errorf("[%s] could not detach channel: %s", localPeer, dc.Label())
+				return
+			}
+			stream = newDataChannel(conn, dc, rwc, pc, nil, nil)
+			select {
+			case acceptQueue <- stream:
+				conn.addStream(id, stream)
+			default:
+				log.Warn("cannot buffer incoming stream, will reset")
+				stream.Reset()
+			}
+
+		})
+
+	})
+
+	pc.OnConnectionStateChange(func(state webrtc.PeerConnectionState) {
+		switch state {
+		case webrtc.PeerConnectionStateDisconnected:
+			log.Warnf("[%s] peer connection disconnected", localPeer)
+		case webrtc.PeerConnectionStateFailed:
+			log.Warn("peer connection reset")
+			conn.resetStreams()
+			fallthrough
+		case webrtc.PeerConnectionStateClosed:
+			log.Warnf("[%s] peer connection closed", localPeer)
+			conn.Close()
+		}
```

Comment:
you do not handle default, yet Go has no guarantees here

Reply:
we do not care about the default case here, and also the switch has been dropped here.